Feature: Add support for FOREIGN KEY

chenxi8611 opened the pull request:
I've closed previous PR and opened new PR.
This PR is the support for FOREIGN KEY related to the https://github.com/duckdb/duckdb/issues/46.

@Mytherin -- Please review it.

Commit:
implement FK

Mytherin commented on src/storage/data_table.cpp:
```diff
@@ -881,19 +1053,49 @@ void DataTable::VerifyUpdateConstraints(TableCatalogEntry &table, DataChunk &chu
+				}
+				if (is_verify_delete) {
+					DataChunk verify_chunk;
+					GatherVerifyChunk(context, row_ids, count, verify_chunk);
```
This chunk, and `GatherVerifyChunk`, appear to never be called. Could you add a test case that tests them?

Mytherin commented on src/planner/binder/statement/bind_update.cpp:
```diff
@@ -105,6 +119,11 @@ static void BindUpdateConstraints(TableCatalogEntry &table, LogicalGet &get, Log
 
+	// if there is a self referencing foreign key constraint, update_is_del_and_insert sets to false
+	if (is_must_update) {
+		update.update_is_del_and_insert = false;
```
I'm still not convinced this is correct - `update_is_del_and_insert` is set to true only when columns are updated that are contained in indexes, and **needs to** be set to true in this case since indexes do not correctly handle updates to individual columns.

Setting this to false when there are columns contained in the index being updated will lead to the index containing outdated or incorrect information, which will lead to wrong query results when the index is used.

Could you add tests that use the index after this condition is being triggered, e.g. through an equality predicate on the indexed column (e.g. `SELECT * FROM tbl WHERE indexed_column=1`).

Mytherin commented on src/planner/binder/statement/bind_export.cpp:
```diff
@@ -39,6 +40,51 @@ string SanitizeExportIdentifier(const string &str) {
+
+void ScanForeignKeyTable(vector<TableCatalogEntry *> &ordered, vector<TableCatalogEntry *> &unordered,
+                         bool move_only_pk_table) {
+	for (vector<TableCatalogEntry *>::iterator i = unordered.begin(); i != unordered.end();) {
```
`auto i`

Mytherin commented on src/storage/data_table.cpp:
```diff
@@ -437,7 +437,112 @@ static void VerifyCheckConstraint(TableCatalogEntry &table, Expression &expr, Da
+		dst_chunk.data[(*dst_keys_ptr)[i]].Reference(chunk.data[(*src_keys_ptr)[i]]);
+	}
+	dst_chunk.SetCardinality(chunk.size());
+	std::shared_ptr<DataTable> data_table = table_entry_ptr->storage;
```
Do we need to take shared pointer ownership here, or would a reference be sufficient (e.g. `auto &data_table = *table_entry_ptr->storage`?

Mytherin commented on src/planner/binder/statement/bind_create.cpp:
```diff
@@ -211,6 +212,33 @@ BoundStatement Binder::Bind(CreateStatement &stmt) {
+		auto &catalog = Catalog::GetCatalog(context);
+		for (idx_t i = 0; i < create_info.constraints.size(); i++) {
+			auto &cond = create_info.constraints[i];
+			if (cond->type == ConstraintType::FOREIGN_KEY) {
```
This is very deeply nested which makes it hard to read. Could we switch to [prefer early return](https://github.com/Shopify/web-configs/blob/main/packages/eslint-plugin/docs/rules/prefer-early-return.md) style here (which is our generally preferred style of coding)?


For example:

```cpp
if (cond->type != ConstraintType::FOREIGN_KEY) {
    continue;
}
if (fk.info.type != ForeignKeyType::FK_TYPE_FOREIGN_KEY_TABLE) {
    continue;
}
...

```

Mytherin commented on src/catalog/catalog_entry/schema_catalog_entry.cpp:
```diff
@@ -31,12 +31,40 @@
+	auto *table_entry = (TableCatalogEntry *)entry;
+	for (idx_t i = 0; i < table_entry->constraints.size(); i++) {
+		auto &cond = table_entry->constraints[i];
+		if (cond->type == ConstraintType::FOREIGN_KEY) {
```
Same here, 

```cpp
if (cond->type != ConstraintType::FOREIGN_KEY) {
    continue;
}
```

Mytherin reviewed (commented):
Thanks for the updates! Looks good. Some remaining comments:

chenxi8611 committed:
Foreignkey (#1)

clean code

chenxi8611 committed:
Merge branch 'duckdb:master' into master

chenxi8611 committed:
Foreignkey (#2)

restore original code

chenxi8611 committed:
Foreignkey (#3)

restore original code

chenxi8611 commented:
@Mytherin -- I've updated the PR as you pointed. Please review it again.

Mytherin reviewed (commented):
Thanks for the updates! This is getting very close now. Some more comments:

chenxi8611 commented on src/storage/data_table.cpp:
```diff
@@ -437,7 +437,112 @@ static void VerifyCheckConstraint(TableCatalogEntry &table, Expression &expr, Da
+		return;
+	}
+
+	vector<string> err_msgs, tran_err_msgs;
```
This array is need because we must check both data table's index and transaction local storage's index.
You can see this by looking at the code below using these arrays.

Mytherin replied:
My point is that this does not need to be an array, it can be a single `string` (`string err_msg`) - since only ever one error message string is used. This should simplify a bunch of the code here, and also make it significantly faster.

chenxi8611 replied:
But we need to look concurrently the error messages in data table's index and transaction local storage's index.
Please look up at the following code.

	for (idx_t i = 0; i < count; i++) {
		if ((!transaction_check && !err_msgs[i].empty()) ||
		    (transaction_check && ((is_append && !err_msgs[i].empty() && !tran_err_msgs[i].empty()) ||
		                           (!is_append && (!err_msgs[i].empty() || !tran_err_msgs[i].empty()))))) {
			string &err_msg = err_msgs[i];
			if (err_msg.empty()) {
				err_msg = tran_err_msgs[i];
			}
			throw ConstraintException(err_msg);
		}
	}

chenxi8611 committed:
Foreignkey (#4)

add multiple test and clean some code

chenxi8611 commented:
@Mytherin -- I've fixed all as you pointed, please review it again.

chenxi8611 committed:
Foreignkey (#5)

fix compile error

chenxi8611 committed:
Foreignkey (#6)

fix fall through error

chenxi8611 committed:
Foreignkey (#7)

change index constructor

chenxi8611 committed:
Foreignkey (#8)

add a test with adding a foreign key reference to a table that has a CHECK constraint

chenxi8611 committed:
Merge branch 'duckdb:master' into master

chenxi8611 committed:
Foreignkey (#9)

update multiple fk test

chenxi8611 commented on src/execution/index/art/art.cpp:
```diff
@@ -877,4 +854,63 @@ bool ART::Scan(Transaction &transaction, DataTable &table, IndexScanState &table
+		Node *node_ptr = Lookup(tree, *keys[i], 0);
+		bool throw_exception =
+		    verify_type == VerifyExistenceType::APPEND_FK ? node_ptr == nullptr : node_ptr != nullptr;
+		if (throw_exception) {
```
Okay, I will clean up as you pointed except one.
     *err_msg_ptr = exception_msg;

This should be changed as the following.
     err_msg_ptr[i] = exception_msg;

chenxi8611 committed:
Foreignkey (#10)

add some comment

Mytherin reviewed (commented):
Thanks for the updates! Some more minor comments. Would like to see a coverage report as well but somehow codecov is acting up again. Perhaps next CI run will succeed there.

chenxi8611 committed:
Foreignkey (#11)

clean code and add comment

chenxi8611 commented:
I've clean the code as you pointed and updated the PR.
Please review it again, thanks

chenxi8611 committed:
Merge branch 'duckdb:master' into master

chenxi8611 committed:
Merge branch 'duckdb:master' into master

Mytherin commented on src/catalog/catalog_entry/schema_catalog_entry.cpp:
```diff
@@ -102,7 +128,28 @@ CatalogEntry *SchemaCatalogEntry::CreateType(ClientContext &context, CreateTypeI
+		if (pk_table_entry_ptr) {
+			info->dependencies.insert(pk_table_entry_ptr);
+		} else {
+			throw ParserException("Can't find table \"%s\" in foreign key constraint", fk_info->name);
```
This code does not seem to be triggered - it also should not be a ParserException. ParserExceptions are special in that they allow a transaction to continue. However, it is possible that you have already made changes to the system here (through the `catalog->Alter`) which can then be (incorrectly) committed. Could you add a test case triggering this and turn this into a `CatalogException`?

chenxi8611 replied:
You are right, this code isn't triggered, I will fix it.

Mytherin commented on src/storage/data_table.cpp:
```diff
@@ -437,7 +437,133 @@ static void VerifyCheckConstraint(TableCatalogEntry &table, Expression &expr, Da
+		// if we are deleting we need to ensure the foreign key DOES NOT exist in EITHER the transaction-local storage
+		// OR the main table
+		if (!err_msgs[i].empty() || !tran_err_msgs[i].empty()) {
+			string &err_msg = err_msgs[i];
```
This is not triggered - and I'm not sure this condition is correct. Could you add a test that triggers this and verify that this is correct?

chenxi8611 replied:
It is triggered in the following case.
CREATE TABLE pkt(i INTEGER PRIMARY KEY);
CREATE TABLE fkt(j INTEGER, FOREIGN KEY (j) REFERENCES pkt(i));
INSERT INTO pkt VALUES (1);
BEGIN TRANSACTION;
INSERT INTO fkt VALUES (1);
DELETE FROM pkt WHERE i=1;

I will add this test.

Mytherin reviewed (commented):
Thanks for the updates! Looks good. Two more remarks and then I think this is good to go:

chenxi8611 committed:
Foreignkey (#13)

fix exception code and add a test

chenxi8611 committed:
Merge branch 'duckdb:master' into master

chenxi8611 committed:
Foreignkey (#14)

restore a deleted comment

Mytherin commented:
Thanks for all the changes! Looks good to me now.

Tishj commented on src/catalog/catalog_entry/schema_catalog_entry.cpp:
```diff
@@ -102,7 +128,23 @@ CatalogEntry *SchemaCatalogEntry::CreateType(ClientContext &context, CreateTypeI
+
+		// make a dependency between this table and referenced table
+		auto &set = GetCatalogSet(CatalogType::TABLE_ENTRY);
+		info->dependencies.insert(set.GetEntry(context, fk_info->name));
```
We passed a copy of `info->dependencies` to AddEntry, changes made to it after this are not reflected in `entry`
